experimental/cuda2/pipeline_layout.c

```diff
@@ -0,0 +1,238 @@
+    iree_hal_descriptor_set_layout_t* const* set_layouts,
+    iree_host_size_t push_constant_count, iree_allocator_t host_allocator,
+    iree_hal_pipeline_layout_t** out_pipeline_layout) {
+  IREE_ASSERT_ARGUMENT(!set_layout_count || set_layouts);
+  IREE_ASSERT_ARGUMENT(out_pipeline_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  *out_pipeline_layout = NULL;
+  if (push_constant_count > IREE_HAL_CUDA_MAX_PUSH_CONSTANT_COUNT) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "push constant count %zu over the limit of %d",
+                            push_constant_count,
+                            IREE_HAL_CUDA_MAX_PUSH_CONSTANT_COUNT);
+  }
+
+  // Currently the pipeline layout doesn't do anything.
+  // TODO: Handle creating the argument layout at that time hadling both push
+  // constant and buffers.
+  iree_hal_cuda2_pipeline_layout_t* pipeline_layout = NULL;
+  iree_host_size_t total_size =
+      sizeof(*pipeline_layout) +
+      set_layout_count * sizeof(iree_hal_descriptor_set_layout_t*);
+  iree_status_t status = iree_allocator_malloc(host_allocator, total_size,
+                                               (void**)&pipeline_layout);
+
+  if (iree_status_is_ok(status)) {
+    iree_hal_resource_initialize(&iree_hal_cuda2_pipeline_layout_vtable,
+                                 &pipeline_layout->resource);
+    pipeline_layout->host_allocator = host_allocator;
+    pipeline_layout->set_layout_count = set_layout_count;
+    iree_host_size_t binding_number = 0;
+    for (iree_host_size_t i = 0; i < set_layout_count; ++i) {
+      pipeline_layout->set_layouts[i] = set_layouts[i];
+      // Copy and retain all descriptor sets so we don't lose them.
+      iree_hal_descriptor_set_layout_retain(set_layouts[i]);
+      binding_number +=
+          iree_hal_cuda2_descriptor_set_layout_binding_count(set_layouts[i]);
+    }
+    pipeline_layout->push_constant_base_index = binding_number;
+    pipeline_layout->push_constant_count = push_constant_count;
+    *out_pipeline_layout = (iree_hal_pipeline_layout_t*)pipeline_layout;
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static void iree_hal_cuda2_pipeline_layout_destroy(
+    iree_hal_pipeline_layout_t* base_pipeline_layout) {
+  iree_hal_cuda2_pipeline_layout_t* pipeline_layout =
+      iree_hal_cuda2_pipeline_layout_cast(base_pipeline_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  for (iree_host_size_t i = 0; i < pipeline_layout->set_layout_count; ++i) {
+    iree_hal_descriptor_set_layout_release(pipeline_layout->set_layouts[i]);
+  }
+  iree_allocator_free(pipeline_layout->host_allocator, pipeline_layout);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_host_size_t iree_hal_cuda2_base_binding_index(
+    iree_hal_pipeline_layout_t* base_pipeline_layout, uint32_t set) {
+  iree_hal_cuda2_pipeline_layout_t* pipeline_layout =
+      iree_hal_cuda2_pipeline_layout_cast(base_pipeline_layout);
+  iree_host_size_t base_binding = 0;
```

Comment:
we could precompute this and store it on the pipeline layout (array of set -> base index) so that the hot path of dispatch is kept simpler

Reply:
Good point! Done.